src/tir/schedule/primitive/pad_einsum.cc

```diff
@@ -0,0 +1,493 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <optional>
+
+#include "../utils.h"
+
+namespace tvm {
+namespace tir {
+
+/*! \brief The schedule error class when the padding size is invalid. */
+class InvalidPaddingError : public ScheduleError {
+ public:
+  InvalidPaddingError(IRModule mod, Block block, Array<Integer> padding)
+      : mod_(std::move(mod)), block_(std::move(block)), padding_(std::move(padding)) {}
+  IRModule mod() const final { return mod_; }
+  Array<ObjectRef> LocationsOfInterest() const final { return {block_}; }
+  String FastErrorString() const final {
+    return "ScheduleError: The padding size for the block is invalid.";
+  }
+  String DetailRenderTemplate() const final {
+    std::ostringstream os;
+    os << "The padding for the block {0} are invalid. It should be a list of "
+       << block_->iter_vars.size() << " non-negative integers. Got " << padding_;
+    return os.str();
+  }
+
+  static void Check(const ScheduleState& self, const Block& block, Array<Integer> padding) {
+    if (padding.size() != block->iter_vars.size()) {
+      throw InvalidPaddingError(self->mod, block, padding);
+    }
+    for (const auto& pad : padding) {
+      if (pad->value < 0) {
+        throw InvalidPaddingError(self->mod, block, padding);
+      }
+    }
+  }
+
+ private:
+  IRModule mod_;
+  Block block_;
+  Array<Integer> padding_;
+};
+
+/*! \brief The schedule error class when the block body is not an Einsum pattern. */
+class NonEinsumError : public ScheduleError {
+ public:
+  explicit NonEinsumError(IRModule mod, Block block)
+      : mod_(std::move(mod)), block_(std::move(block)) {}
+
+  IRModule mod() const final { return mod_; }
+  Array<ObjectRef> LocationsOfInterest() const final { return {block_}; }
+  String FastErrorString() const final {
+    return "ScheduleError: The block is not a computation of Einsum pattern.";
+  }
+  String DetailRenderTemplate() const final {
+    return "The block {0} not a computation of Einsum pattern.";
+  }
+
+ private:
+  IRModule mod_;
+  Block block_;
+};
+
+/*! \brief Data structure that represents a Einsum computation. */
+struct Einsum {
+  // The output buffer
+  Buffer output_buffer;
+  // The indices of the output buffer
+  Array<Var> output_indices;
+  // The indices of the input buffers
+  Map<Buffer, Array<Var>> input_indices;
+};
+
+/*!
+ * \brief Check if buffer indices are all Vars and extr
+ * \param buffer_access The BufferLoad or BufferStore
+ * \param[out] indices The optional array to store the indices
+ * \return Whether The indices are all Vars
+ */
+template <typename T>
+bool CheckTrivialBufferIndices(const T& buffer_access, Array<Var>* indices = nullptr) {
```

Comment:
Instead of an output parameter, could we return either `Optional<Array<Var>>` or `std::optional<Array<Var>>`?  It looks like all the current usages use the extracted indices, so it wouldn't introduce an unnecessary allocation.